New Functionality & Concepts: Child Directories and Clean-Up

ndunn990 opened the pull request:
## Summary

I've added three new commands and two new configuration items that can be included in a tfgen configuration file. I also updated the README to reflect the changes here.

## UPDATE

- I have changed `child_directories` to `target_directories` to be more consistent and better prepared for the future.

## Minor Changes to Config Logic:

- In order to facilitate more flexibility around some of the standard operations tfgen does, I had to add a function or two under config and finder.

## Configuration Items:

- A configuration can now include `child_directories`. This is a list of relative paths from the directory that contains the root configuration. This can be used by the `exec_all` command.
- A configuration can now include a `clean_pattern`. This is a pattern that can be defined for the `clean` and `clean_all` commands.

## Commands:

- `exec_all`: With this command you specify a directory a root configuration lives. As long as it _is_ a root configuration, tfgen will now loop through each directory listed in `child_directories` (if any) and `exec` as normal. This allows you to interact with your entire project at once, assuming there's a hierarchy in place.
- `clean`: With this command you can clean up any files generated via template by tfgen. To allow the user as much control over this as possible, your configuration will need to include a `clean_pattern`, which should just be a simple regex string.
- `clean_all`: Much like the `clean` command. However, this command should target a directory a root configuration lives. As long as it _is_ a root configuration, tfgen will now loop through each directory listed in `child_directories` (if any) and `clean` as normal.

ndunn990 committed:
Config Struct Update

Updated the config struct to include some new properties that will allow
a root configuration to list 'child' directories and a 'clean_pattern'
which will give users some control over how tfgen will clean up templates.

ndunn990 committed:
Finder Update: New Functions for New Commands

Removed some of the logic in the function that searches parent
directories. Moved that logic to its own function ('collateConfig') so
it can be used by other functions, including the ones needed for the new
command 'exec_all'.

ndunn990 committed:
New Commands: 'clean' and 'clean_all'

New commands that will allow tfgent to delete files that were generated
by a template. These commands expect a 'clean_pattern' to be defined,
ideally in the root configuration to allow for the 'clean_all' command,
which will clean the 'child_directories' defined in the root confing.
This should give the user a great deal of control around how they want
things cleaned up when needed.

ndunn990 committed:
New Command: 'exec_all'

Thew 'exec_all' command allows users to specify a directory where a root
configuration is (this is required -- it errors out, otherwise). Then,
tfgen will 'exec' as normal for any and all children directories that
are defined by the root config.

ndunn990 committed:
Adding Commands to Main

Adding the new commands ('exec_all', 'clean', and 'clean_all') to main.

ndunn990 committed:
Updated README with New Commands & Functionality

linuxfreakus commented on config/finder.go:
```diff
@@ -63,46 +63,106 @@ func GetConfigFiles(targetDir string) ([]Config, error) {
-			return configFileRelativePath, emptyMap, nil
-		}
+// Examine the TFGEN configuration situation in a specific directory
+func collateConfig(targetDir, configFileName, configDirName string) (string, string, map[string]string, error) {
```
this is useful, having the logic here extracted into a separate function

linuxfreakus commented on config/config.go:
```diff
@@ -16,6 +16,8 @@ type Config struct {
 	Dependencies          *Dependencies        `yaml:"deps"`
 	Variables             map[string]string    `yaml:"vars"`
 	TemplateFiles         map[string]string    `yaml:"template_files"`
+	ChildDirectories      []string             `yaml:"child_directories"`
```
I am curious about future possibilities of allowing sub directories under a  target/project which also contain generated code.  Not sure what that would look like yet, but this name might conflict with future possibilities there.

Perhaps is should be named `project_directories` or `target_directories`?  I kind of like project but not sure....

ndunn990 replied:
I believe I see what you mean. I'll update them to `target_directories` since I think that's more in-line with the default `targetDir`.

linuxfreakus commented on config/finder.go:
```diff
@@ -63,46 +63,106 @@ func GetConfigFiles(targetDir string) ([]Config, error) {
+	configDirMissingConfig := onlyConfigDirFound && configDirConfigFileRelativePath == ""
+
+	if bothConfigTypesFound {
+		return "BothConfigTypesFound", "", emptyMap, fmt.Errorf("in %s you must use either a config file or config directory but not both", currentDir)
```
It occurs to me that these return values for config status might be nice to have as an enum or some other pattern where if they were misused somehow accidentally, it would show up easier with compilation errors etc.

ndunn990 replied:
Agreed. I was under the impression enums did not exist in Go, but after some digging I've seen there are "official" ways to implement them. I think there would be some value updating these statuses accordingly at some point.

linuxfreakus commented on main.go:
```diff
@@ -28,6 +28,9 @@ func main() {
 	}
 	rootCmd.PersistentFlags().StringVarP(&logLevel, "log-level", "l", "info", "configures the logging level")
 	rootCmd.AddCommand(cmd.NewExecCmd())
+	rootCmd.AddCommand(cmd.NewExecAllCmd())
```
Just noting that in theory exec cold have an argument `--all` and same with clean... but I'm not sure it matters, but it would reduce the number of top level commands.

ndunn990 replied:
I agree. I decided to adhere to Terragrunt's model, which has entirely new/separate top level commands when executing for multiple levels within a hierarchy. I did the same since tfgen is meant to be a sort of alternative. However, I think we could easily move it to a `--all` flag if we decide it makes more sense or if tfgen starts getting a bit noisy/busy.

linuxfreakus reviewed (changes requested):
I had some comments inline, for the most part I'm a little hesitant about the `child_directories` naming, considering possible future uses, most of the other stuff are observations which could be worth a little discussion.

ndunn990 committed:
Child Directories -> Target Directories

The term 'child_directories' might become confusing later on due to some
other features that might be implemented. Going to change them to
'target_directories' which is a bit more in line with the original
default behavior of tfgen and how it named/identified the 'targetDir'.

ndunn990 commented:
> I had some comments inline, for the most part I'm a little hesitant about the `child_directories` naming, considering possible future uses, most of the other stuff are observations which could be worth a little discussion.

I think I understand what you're thinking for the potential future. So, I've decided to switch their name to `target_directories`. I feel that's more in line with the original default behavior since it would refer to the directory being executed as the 'target' directory.

linuxfreakus reviewed (approved):
I'll approve this, I think it adds some very useful changes and keeps the code changes fairly small still since I don't know if the original author I forked this from might eventually also want to jump back in and get more involved too.

I would definitely like to keep this all pretty light touch and mostly a helper tool not a big swiss army knife :P
